### PR TITLE
Add ability to specify custom headers

### PR DESCRIPTION
**Pull request by kceb**
Add ability to add custom headers to the endpoint prop

Use case: I want to be able to set the CSRF-TOKEN header instead of using the meta tag approach.

https://github.com/cbothner/react-activestorage-provider/issues/9

### CODE REVIEW

**Comment by cbothner** [comment]
Thanks so much for the PR. Sorry I took a while to respond. I was in the middle of a bit of a refactor (#11) that I wanted to be able to reference.

First of all, yes absolutely. Great idea. I think accepting an object of headers is a great general solution, better than special casing like we did with `Authorization` in #8. 

Let’s make `headers` an optional prop on `ActiveStorageProvider` itself, rather than being a key in `endpoint`, since it’s not really a part of the URI. Its type should be `{[key: string]: string}` [so that it’s compatible with `fetch`](https://github.com/facebook/flow/blob/b448d8d9b7e7918825f925cfbf7294ce79530386/lib/bom.js#L924). I would like to remove the `token` prop in favor of `headers`, but I can do that in a separate PR if you’d like.

Then, I think we should pass the headers prop through the new `DirectUploadsProvider` and into `Upload` so that we can include them with the blob request too. I think that makes sense, since it might be necessary if someone has overridden `ActiveStorage::DirectUploadsController` to implement auth, or the like. The `activestorage` package gives us the opportunity for us to modify the XHR before the request is sent, so we can loop through the headers object and `xhr.setRequestHeader` in `directUploadWillCreateBlobWithXHR`.

And last, if someone has set the `X-CSRF-Token` header, we don’t need to look for the meta tag / throw an error if it’s not found. Can we check for that header’s inclusion in `headers` and only conditionally call `csrfHeader()`?

**Comment by kceb** [comment]
@cbothner good feedback, will address. I agree with everything that you said. Thanks!

File: src/Upload.test.js [hunk 1]
```diff
@@ -127,6 +127,24 @@ describe('Upload', () => {
           xhr,
         })
       })
+
+      describe('if headers are provided', () => {
+        const mockXHR = {
+          setRequestHeader: jest.fn(),
+          readyState: 1,
```
Comment (kceb) [inline src/Upload.test.js]: I'm actually not sure if the xhr object is in readyState 1 (opened) when `directUploadWillCreateBlobWithXHR` is called. Ideas?

Comment (cbothner) [reply src/Upload.test.js]: The `activestorage` package makes heavy use of `setRequestHeader` before we get the `xhr`, so it should always be ready to go. https://github.com/rails/rails/blob/5-2-1/activestorage/app/javascript/activestorage/blob_record.js#L15

Checking with a breakpoint in my app, it seems to work.

File: src/index.js [hunk 2]
```diff
@@ -76,7 +78,8 @@ class ActiveStorageProvider extends React.Component<Props> {
         Accept: 'application/json',
         'Content-Type': 'application/json',
         ...(token ? { Authorization: token } : {}),
-        ...csrfHeader(),
+        ...addCSRFHeader && csrfHeader(),
```
Comment (cbothner) [inline src/index.js]: Flow doesn’t like when you spread booleans into objects, even though it works.

```suggestion
        ...(addCSRFHeader ? csrfHeader() : {}),
```

**Comment by ivansifrim** [comment]
This looks great. Can't wait to try this feature.

**Comment by kceb** [comment]
@cbothner much appreciated!

For some reason when testing in my browser, it prepends `undefined` for the X-CSRF-Token header: 
https://kceb.d.pr/jBg7MP

Is this being set on the rails side? I tried clearing it then setting it locally and it still prepends the value with `undefined, [CSRF_TOKEN_VALUE]`

**Comment by cbothner** [comment]
Finally got around to submitting a PR and it got merged. Should be fixed in Rails 6.0

**Pull request merged**